src/templates.c

```diff
@@ -255,19 +255,57 @@ static void add_file_item(const gchar *fname, GtkWidget *menu)
 }
 
 
+typedef struct
+{
+	gint count;
+	GtkWidget *menu;
+}
+FTMenu;
+
 static void populate_file_template_menu(GtkWidget *menu)
 {
 	GSList *list = utils_get_config_files(GEANY_TEMPLATES_SUBDIR G_DIR_SEPARATOR_S "files");
 	GSList *node;
+	FTMenu *ft_groups;
+	gint nbytes = sizeof(FTMenu) * filetypes_array->len;
+
+	ft_groups = g_alloca(nbytes);
+	memset(ft_groups, 0, nbytes);
+
+	foreach_slist(node, list)
+	{
+		gchar *fname = node->data;
+		GeanyFiletype *ft = filetypes_detect_from_extension(fname);
 
+		if (ft)
+			ft_groups[ft->id].count++;
+	}
 	foreach_slist(node, list)
 	{
 		gchar *fname = node->data;
+		GeanyFiletype *ft = filetypes_detect_from_extension(fname);
+		FTMenu *group = NULL;
 
-		add_file_item(fname, menu);
+		if (ft)
+			group = &ft_groups[ft->id];
+		if (group->count == 1)
```

Comment:
`group` may be NULL if a file in the directory does not have a filetype extension

Reply:
I realized `ft` is never NULL in this function. The `if (ft)` checks were redundant, removed.

Reply:
Ahh, ok, `filetypes_detect_from_extension()` returns filetype none if it doesn't recognise it.